File: Data Structures & Algorithms/subsets-ii/submission-2.py
class Solution:
    def subsetsWithDup(self, nums: List[int]) -> List[List[int]]:
        

        res = set()

        def backtrack(i, subset):
            
            # Base Case
            if i == len(nums):
                res.add(tuple(subset))
                return

            # Recursive Case
            subset.append(nums[i])
            backtrack(i+1, subset)

            subset.pop()
            backtrack(i+1, subset)

        nums.sort()
        backtrack(0, [])
        return [list(s) for s in res]

File: Data Structures & Algorithms/combination-target-sum/submission-2.java
class Solution {

    public List<List<Integer>> combinationSum(int[] nums, int target) {
        
        Arrays.sort(nums);
        List<Integer> cur = new ArrayList<>();
        List<List<Integer>> res = new ArrayList<>();
        backtrack(nums, target, cur, 0, res);
        return res;
    }

    private void backtrack(int[] nums, int target, List<Integer> cur, int i, List<List<Integer>> res) {

        // Base Case
        if ( target == 0 ) {
            res.add(new ArrayList<>(cur));
            return;
        }

        // Base Case 2
        if (target < 0 || i >= nums.length) return;

        cur.add(nums[i]);
        backtrack(nums, target - nums[i], cur, i, res);
        cur.remove(cur.size() - 1);
        backtrack(nums, target, cur, i + 1, res);
    }
}
